### ISSUE TITLE
Make all dialog windows set to always on top

### ISSUE THREAD

**Issue by audio2000** [lsp-plugins/lsp-plugins#440]
Linux

If possible maybe can make all dialog windows set to be always on top.
Dialog windows such as.....
-File Browser / Pickers
-“About” dialog
-"User Paths" dialog (used in sampler plugins)
-Etc.

I noticed that these dialogs can get sent to the back behind existing opened windows, becoming not visible, sometimes making the plugin seem like it is not working but really there is just a dialog window open in the background preventing access to the plugin GUI itself.
Not sure if anyone else has ever experienced this.

Thank You

**Comment by sadko4u** [comment]
Do you use latest Linux build? AFAIR I did some code to bring the windows to top. Can you give any reproducible case for that?

**Comment by audio2000** [comment]
Thank you for responding, I am experiencing this in
Package Version: 1.2.16 LV2
Do you know if the changes you made were done in the latest version you have in your release section, currently 1.2.17

Even more specifically I tried this out using "Compressor" Mono plugin (1.0.24)
I experienced this in Ardour 8.6.0 and this occurs in newest release 8.7.0 also.
If you open the plugins "About" window and you click a window behind it, the About window then goes all the way behind Ardour window itself so you do not know it is opened.  You have to minimize Ardour to close the About window behind it.
I also 
Ardour - Preferences - Quirks I believe this the only settings available to adjust how windows are handled.
I think I have the default settings set where only the top two items are checked on.
I also tried this with Sampler Mono so I can test this with a File Picker window which same thing occurs.

Upon further testing, I tried this out in Carla and this DOES work properly in this application.
The Above & File Picker windows stay "on top" making the plugin behind it not able to receive any adjusting, which I believe is all the expected behavior.

If this occurs for you as well in Ardour, then maybe you can contact them specifically to ask why?  (If you cannot find the issue in your plugins themselves)

Hopefully some of this information is helpful.
Thank You

**Comment by sadko4u** [comment]
> Do you know if the changes you made were done in the latest version you have in your release section, currently 1.2.17

Yes, the changes were made in the recent one. So I would like to recommend to update and check the behaviour again.

Ardour is too specific to plugin windows. I'm already tired fighting with it regarding the control of window size.

**Comment by audio2000** [comment]
Okay, I have tested this out again using the newer version from this site 1.2.17 and experienced the same results.
I tried this in more than one desktop environment just to make sure it was not DE / WM specific.
Carla = works fine
Ardour 8.7.0 = does not, the About and or File Pickers go behind Ardour itself if you click on LSP plugin or Ardour itself.
Not sure if anyone else can verify?
Thank You

**Comment by sadko4u** [comment]
OK, I need to play with X11 protocol a bit more then...

**Comment by audio2000** [comment]
Thank you for looking into this.
Like I said it might just be me, if you or no one else experiences this?
As mentioned only occurs in Ardour, this did not occur in Carla.
Just thought I would mention because it could confuse other users in the future if this ever happens to them.
They might look at LSP plugins to have bugs that "freeze" plugins, but really it is just a dialog window might be in background.
Thank you again

**Comment by sadko4u** [comment]
Please check the 1.2.18 release if you have problems with windows.

**Comment by audio2000** [comment]
Just reporting that I tried this out in version 1.2.19 release and still occurs.
Again it could just be me if you and other users are not experiencing this.
I will try and see if I can ask around and see if I can gather any more information about why this might be occurring.
As has been mentioned Ardour does seem to have some issues with windows.
This does not occur when I test this out in Carla with LSP plugins.
So it is either something with how Ardour interacts with LSP plugins or it is something with Ardour itself?
I mean I am aware when/if this occurs, so I know to correct it by hitting Super+D to show desktop and to close the window, but others might not.
I just didn't want them to look at the plugins you guys create and think there are issues with them freezing or anything like that, because when this occurs that is what it could appear to user as.
Anyways, I will see if I can get some more information but not important if cannot be fixed, up to you.
Thank you again

**Comment by audio2000** [comment]
Sorry I have been trying to generate more info about this issue, seeing if others can share feedback but no luck yet.
This unfortunately still occurs when using Ardour.
It has been mentioned that Carla doesn't set any preferences to plugins and its child windows, but Ardour does.
If you come across anything about this issue please let me know.
Thank You

**Comment by sadko4u** [comment]
I don't have any idea. I suppose it is heavy dependent on the window manager. In my KDE environment dialogs are always above the plugin's window. I also tested with other WMs and got the same result.
I suppose I did here all what I can.

**Comment by audio2000** [comment]
@sadko4u
Up to you if at all interested but I mentioned I would try to help in any way that I could to narrow down any issues found.

I came back to this post to share something that I recently noticed, that actually correlates with what I just read in your response above.

This might be related to something with Linux GTK based applications.


---------------------------------------------------------

`Qt BASED APPLICATIONS`
Carla
I had mentioned that this issue did not occur with the application Carla, at the time I did not realize that Carla used Qt toolkit.

KDE Plasma
You mentioned this did not occur for you in KDE environment, which is also Qt based

QTractor
QTractor is a little strange because it does not embed plugin GUI windows within the QTractor window itself.
So if you click on the QTractor window, LSP plugin will go behind.
However this does work in the sense that the LSP windows such as
`Open Dialog` window / `About` window 
do not go behind the LSP plugin window.

---------------------------------------------------------

`GTK BASED APPLICATIONS`

Ardour
This application is GTK based and issue occurs

Reaper
I was trying to find any GTK based DAW applications I could that can open plugins.  I recently tried out Reaper (which I believe is GTK based?) and the same issue does occur.

---------------------------------------------------------

So it is not Ardour specific, it might be a GTK specific issue?

Now you mentioned you tried other WMs and did not experience this?
I assume that any other WMs you tried would be GTK based so maybe my GTK based assumption is still not true?

You closed this issue so I cannot reopen it myself.
So again, up to you if still interested in fixing something like this / reopening this post.

Anyways just thought I would share this observation.

Thank You

**Comment by sadko4u** [comment]
REAPER is not GTK-based. So your assumption about GTK-related problems is not correct.
What DE did you use to perform your tests? Have you tried another DE?

**Comment by audio2000** [comment]
@sadko4u
Okay I have put even more time into trying to provide as much information about this issue as I possibly can.

As mentioned at that time I could not find what toolkit REAPER is using?
I looked it up again and for some reason this time I was able to generate an answer.
I found that it might be using something called `WDL` (Cockos' WDL) toolkit?
If this is true, then my assumption may be correct.
Because in this case, then it does not have any connection to `Qt` which I might have something to do with this.

I have tested out other Linux based DEs
Ardour 9.5.x
LSP 1.2.29 LV2

`GTK BASED DEs`
Cinnamon
MATE
XFCE
All of these produced the exact same issue, floating LSP dialogs/windows get sent all the way to the back, behind all application windows.

`Qt BASED DE`
KDE Plasma
This issue does not occur at all, works perfectly.
Which may be why you are not experiencing this?

So my theory of, any time there is a `Qt` based DE or `Qt` based application used in the test, this issue does not occur.
It seems like any kind of connection to `Qt` then this issue does not occur.

The only thing that may prove this wrong is something else you mentioned, that you did not experience this using some WMs.   The only WM I know of that uses `Qt` is KWin, but if you used any others that have no connection to `Qt` then as mentioned my theory may not be correct.

Either way though, it does occur in all the DEs tested other then Qt, so the issue is not linked to a specific DE / DE based window manager, it is something else?
Not sure if this helps narrow anything down in any way?

Thank You

**Comment by audio2000** [comment]
@sadko4u 
Hello again
I do not have the ability to reopen this post, but I was wondering if it would be okay with you if I created a new post for this topic now that new information has been generated/mentioned above?
Thank You

**Comment by sadko4u** [comment]
I'll just reopen the issue and have one more try to reproduce/fix it.
